src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java

```diff
@@ -0,0 +1,238 @@
+    consumerProps.put(ConsumerConfig.CLIENT_ID_CONFIG, KMF_CONSUMER);
+    consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, KMF_CONSUMER_GROUP_PREFIX + new Random().nextInt());
+    consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+    consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+    consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList);
+    consumerProps.put(CommonServiceConfig.ZOOKEEPER_CONNECT_CONFIG, zkConnect);
+    if (consumerClassName.equals(NewConsumer.class.getSimpleName())) {
+      consumerClassName = NewConsumer.class.getCanonicalName();
+    }
+    Map consumerPropsOverride;
+    if (props.containsKey(ConsumeServiceConfig.CONSUMER_PROPS_CONFIG)) {
+      consumerPropsOverride = (Map) props.get(ConsumeServiceConfig.CONSUMER_PROPS_CONFIG);
+    } else {
+      consumerPropsOverride = new HashMap<>();
+    }
+    consumerProps.putAll(consumerPropsOverride);
+    if (props.containsKey(ConsumeServiceConfig.CONSUMER_PROPS_CONFIG)) {
+      props.forEach(consumerProps::putIfAbsent);
+    }
+  }
+
+  private void initiateCommit() throws Exception {
+    LOG.info("Commit starting in {}.", getClass().getSimpleName());
+    try {
+      Thread.sleep(THREAD_SLEEP_MS);
+    } catch (InterruptedException exception) {
+      LOG.error("Error occurred while sleeping the thread ", exception);
+    }
+
+    while (_running.get()) {
+      BaseConsumerRecord baseConsumerRecord;
+      try {
+        baseConsumerRecord = _kmBaseConsumer.receive();
+      } catch (Exception exception) {
+        LOG.warn(_name + "/ConsumeService failed to receive record.", exception);
+        /* Avoid busy while loop */
+        try {
+          long threadSleepMs = 100;
+          Thread.sleep(threadSleepMs);
+        } catch (InterruptedException ex) {
+          LOG.error("Interrupted Exception occurred while trying to sleep the thread.", ex);
+        }
+        continue;
+      }
+      if (baseConsumerRecord == null) continue;
+      GenericRecord genericAvroRecord = Utils.genericRecordFromJson(baseConsumerRecord.value());
+      int partition = baseConsumerRecord.partition();
+      long index = (Long) genericAvroRecord.get(DefaultTopicSchema.INDEX_FIELD.name());
+      if (index == -1L || !_nextIndexes.containsKey(partition)) {
+        _nextIndexes.put(partition, -1L);
+        continue;
+      }
+
+      try {
+        TopicPartition topicPartition = new TopicPartition(_topic, partition);
+        commitAndRetrieveOffsets(topicPartition, _offsetsToCommit);
+
+      } catch (KafkaException kafkaException) {
+        LOG.error("Exception while trying to to async commit.", kafkaException);
+        _commitAvailabilityMetrics._failedCommitOffsets.record();
+      }
+    }
+  }
+
+  private OffsetAndMetadata commitAndRetrieveOffsets(TopicPartition topicPartition, Map<TopicPartition, OffsetAndMetadata> offsetMap) throws Exception {
```

Comment:
Since it’s impossible to add testing for commitRetrieveOffsets without initializing a new CommitAvailabilityService instance in the Test  class. Majority of KMF code is badly structured (when originally written since its inception) for unit testing.